impeller/geometry/rect_unittests.cc

```diff
@@ -44,16 +44,200 @@ TEST(RectTest, RectMakeSize) {
     Size s(100, 200);
     IRect r = IRect::MakeSize(s);
     IRect expected = IRect::MakeLTRB(0, 0, 100, 200);
-    ASSERT_EQ(r, expected);
+    EXPECT_EQ(r, expected);
   }
 
   {
     ISize s(100, 200);
     IRect r = IRect::MakeSize(s);
     IRect expected = IRect::MakeLTRB(0, 0, 100, 200);
-    ASSERT_EQ(r, expected);
+    EXPECT_EQ(r, expected);
   }
 }
 
+TEST(RectTest, NormalizePoint) {
```

Comment:
I'd rather see individual TEST cases versus a single case with a list of things it tests (you can pull up/re-use the `test_one` closure I suppose). Basically any place you have to explain what different things are being tested.

Reply:
There are 630+ cases tested there. Do you want 630 TEST macros? If we require a TEST invocation for each and every condition tested we are incentivizing spot testing rather than rigorous testing.

Reply:
I'm just communicating I wouldn't understand what "failure: NormalizePoint line 969" means in a test log. Maybe that means a few (not 630, but not 3) test macros, and maybe it doesn't.

Reply:
I guess I'm curious as to how much factorization you see being appropriate here. In the long view they are testing just a single Rect method. In the near view there's a lot of combinatorics and breaking up each combinatoric case would be unwieldy. But, maybe there is a slice in all of the cases that can be broken out? I'd want to avoid:

```
TEST(RectTest, NormalizePointWhereRectHasNaNXOriginAndPointHasInfiniteYOnRectAndIPoint) {
  ...
}
```

Reply:
> I'm just communicating I wouldn't understand what "failure: NormalizePoint line 969" means in a test log. Maybe that means a few (not 630, but not 3) test macros, and maybe it doesn't.

Which is why I use the labeling feature. The output contains the cases being tested.

Reply:
I thought of a better way to use the labeling feature. Right now I'm using those strings to try to print out an "english description" of what the test case was, but perhaps I should just output `Test Case: [I]Rect::MakeLTRB(<the values>).NormalizePoint([I]Point(<the values>)`? No need to explain that this is the "UL" or "Upper Left Corner of the Source Rectangle" - just let the values do the explaining...

Reply:
Basically, use this to run every test case so that the failures are self-documenting:

```
  // Runs a single test case and outputs the parameters used on failure
  // to make it easier to diagnose which test case failed.
  auto test_case = [](Rect r, Point p, Point expected) {
    EXPECT_EQ(r.NormalizePoint(p), expected)
        << "Rect" << r << ".NormalizePoint(Point" << p << ")";
  };
```

Reply:
I ran with a variant of this using a macro that will perform the test and then print out the data on failure which helps identify which values are not compliant with the test/spec.

I didn't factor out the 2 cases any more than they had been, though, but they should be easier to read and the error output will be self-revealing.

Reply:
I sabotaged one of the tests to get some error output to show what it looks like:

```
../../flutter/impeller/geometry/rect_unittests.cc:80: Failure
Expected equality of these values:
  Rect::MakeLTRB(l, t, r, b).NormalizePoint(Point(px, px))
    Which is: (1, 1)
  expected
    Which is: (1, 0)
Rect::MakeLTRB(100, 100, 200, 200).NormalizePoint(Point(200, 200))

../../flutter/impeller/geometry/rect_unittests.cc:80: Failure
Expected equality of these values:
  Rect::MakeLTRB(l, t, r, b).NormalizePoint(Point(px, px))
    Which is: (0, 0)
  expected
    Which is: (0, 1)
Rect::MakeLTRB(100, 100, 200, 200).NormalizePoint(Point(100, 100))

../../flutter/impeller/geometry/rect_unittests.cc:80: Failure
Expected equality of these values:
  Rect::MakeLTRB(l, t, r, b).NormalizePoint(Point(px, px))
    Which is: (2, 2)
  expected
    Which is: (2, -1)
Rect::MakeLTRB(100, 100, 200, 200).NormalizePoint(Point(300, 300))
```